grpc-api/src/main/java/io/deephaven/grpc_api/table/ops/CreateInputTableGrpcImpl.java

```diff
@@ -0,0 +1,84 @@
+
+import com.google.rpc.Code;
+import io.deephaven.datastructures.util.CollectionUtil;
+import io.deephaven.db.tables.Table;
+import io.deephaven.db.tables.TableDefinition;
+import io.deephaven.db.v2.utils.AppendOnlyArrayBackedMutableTable;
+import io.deephaven.db.v2.utils.KeyedArrayBackedMutableTable;
+import io.deephaven.extensions.barrage.util.BarrageUtil;
+import io.deephaven.extensions.barrage.util.GrpcUtil;
+import io.deephaven.grpc_api.session.SessionState;
+import io.deephaven.proto.backplane.grpc.BatchTableRequest;
+import io.deephaven.proto.backplane.grpc.CreateInputTableRequest;
+import io.grpc.StatusRuntimeException;
+import org.apache.arrow.flatbuf.Message;
+import org.apache.arrow.flatbuf.MessageHeader;
+import org.apache.arrow.flatbuf.Schema;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Collections;
+import java.util.List;
+
+import static io.deephaven.proto.backplane.grpc.CreateInputTableRequest.InputTableKind.KindCase.KIND_NOT_SET;
+
+@Singleton
+public class CreateInputTableGrpcImpl extends GrpcTableOperation<CreateInputTableRequest> {
+
+    private static final MultiDependencyFunction<CreateInputTableRequest> optionalSourceTable =
+            (CreateInputTableRequest req) -> req.hasSourceTableId()
+                    ? Collections.singletonList(req.getSourceTableId())
+                    : Collections.emptyList();
+
+    @Inject
+    public CreateInputTableGrpcImpl() {
+        super(BatchTableRequest.Operation::getCreateInputTable, CreateInputTableRequest::getResultId,
+                optionalSourceTable);
+    }
+
+    @Override
+    public void validateRequest(CreateInputTableRequest request) throws StatusRuntimeException {
+        // ensure we have one of either schema or source table (protobuf will ensure we don't have both)
+        if (!request.hasSchema() && !request.hasSourceTableId()) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                    "Must specify one of schema and source_table_id");
+        }
+
+        if (request.getKind().getKindCase() == null ||
+                request.getKind().getKindCase() == KIND_NOT_SET) {
+            throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                    "Unrecognized InputTableKind");
+        }
+    }
+
+    @Override
+    public Table create(CreateInputTableRequest request, List<SessionState.ExportObject<Table>> sourceTables) {
+        TableDefinition tableDefinitionFromSchema;
+
+        if (request.hasSchema()) {
+            Message message = Message.getRootAsMessage(request.getSchema().asReadOnlyByteBuffer());
+            if (message.headerType() != MessageHeader.Schema) {
+                throw GrpcUtil.statusRuntimeException(Code.INVALID_ARGUMENT,
+                        "Must specify schema header in schema message");
+            }
+            tableDefinitionFromSchema = BarrageUtil.convertArrowSchema((Schema) message.header(new Schema())).tableDef;
```

Comment:
Looks like from this context we don't have to worry about the magic 8 bytes?

I wonder if we should try to use the utility code, or generalize the utility code, in io.deephaven.grpc_api.util.SchemaHelper and/or BarrageUtil instead of doing it inline here?

Reply:
Added a new method to SchemaHelper to unwrap an arbitrary ByteBuffer, then pass to BarrageUtil to convert to the TableDefinition.